server/src/main/java/org/elasticsearch/index/reindex/ReindexRequest.java

```diff
@@ -100,11 +98,9 @@ public ActionRequestValidationException validate() {
         if (false == routingIsValid()) {
             e = addValidationError("routing must be unset, [keep], [discard] or [=<some new value>]", e);
         }
-        if (destination.versionType() == INTERNAL) {
```

Comment:
this is stronger than what `validateVersionAndSeqNoBasedCASParams` does. It basically says - if you use internal versioning, you shouldn't set a specific version. I think that's good? also, we probably want the same for external versioning (i.e., the version it self can't be set) and something ifSeqNo/ifPrimary term (which I missed).

Reply:
Note that the next line here says:

```
if (destination.version() != Versions.MATCH_ANY && destination.version() != Versions.MATCH_DELETED) {
```

and `validateVersionAndSeqNoBasedCASParams` already has

```
if (versionType == VersionType.INTERNAL && version != Versions.MATCH_ANY && version != Versions.MATCH_DELETED) {
    validationException = addValidationError("internal versioning can not be used for optimistic concurrency control. " +
        "Please use `if_seq_no` and `if_primary_term` instead", validationException);
}
```

These look equivalent to me?

Reply:
> and `validateVersionAndSeqNoBasedCASParams` already has

fair enough (those were folded away and I missed them). Also that would have meant this can't go to 6.x 🤷‍♂️